### PR TITLE
collect in cluster/array descriptions

### PR DESCRIPTION
**Pull request by burrbull**
replaces #177
cc @kossnikita

### CODE REVIEW

**Review by kossnikita — CHANGES_REQUESTED** [review]
Fail with

```yaml
_svd: stm32l4x2.svd
LCD:
  RAM_COM6:
    _array:
      S?? : {}
```

```
RAM_COM6: fields cannot be collected into S?? array. Please, specify description
```

File: src/patch/mod.rs [hunk 1]
```diff
@@ -637,3 +638,44 @@ fn check_offsets(offsets: &[u32], dim_increment: u32) -> bool {
...
+
+fn first_diff(s1: &str, s2: &str) -> usize {
+    for ((i, c1), c2) in s1.char_indices().zip(s2.chars()) {
+        if c1 != c2 {
+            return i;
+        }
+    }
+    s1.len().min(s2.len())
+}
+fn replacefirstfrom<'a>(s: &'a str, pat: &str, to: &str, i: usize) -> Cow<'a, str> {
```
[the remark this answers is not in the document]
Comment (kossnikita) [reply src/patch/mod.rs:650]: could you also add short descriptions of these functions?

**Comment by burrbull** [comment]
cc @Emilgardis

**Comment by kossnikita** [comment]
What about https://github.com/rust-embedded/svdtools/pull/178#pullrequestreview-1705585405?
I don't like that this breaks existing patches that previously worked correctly

**Comment by burrbull** [comment]
> What about [#178 (review)](https://github.com/rust-embedded/svdtools/pull/178#pullrequestreview-1705585405)? I don't like that this breaks existing patches that previously worked correctly

Sorry missed this comment.
Can't reproduce. Is it `stm32-rs` repo? Could you retry on last `stm32-rs` master and this PR?

**Comment by burrbull** [comment]
Saying about "RAM_COMx". It looks like there are missed high parts of "64-bit" registers.

**Comment by kossnikita** [comment]
> Is it `stm32-rs` repo?

This is 
https://github.com/rust-embedded/svdtools/blob/master/example/stm32l4x2.svd

**Comment by burrbull** [comment]
> > Is it `stm32-rs` repo?
> 
> This is [svdtools/example/stm32l4x2.svd](https://github.com/rust-embedded/svdtools/blob/master/example/stm32l4x2.svd) My example about auto collecting discriptions like `S??`. Before PR this works.

Thanks. I totally forgot about this example.
I think this is one of special cases where we need to specify description manually. (or maybe even `collect` should not work at all).
The issue is names of these fields. They are looks like numerical (and can be parsed as numerical) but instead of `S0, S1, S2, .., S31` they are `S01, S02, S03, .., S31`. So before this patch result looks like (both `description` and `dimIndex` are incorrect):
```xml
        <register>
          <name>RAM_COM6</name>
          <displayName>RAM_COM6</displayName>
          <description>display memory</description>
          <addressOffset>0x44</addressOffset>
          <size>0x20</size>
          <access>read-write</access>
          <resetValue>0x00000000</resetValue>
          <fields>
            <field>
              <dim>32</dim>
              <dimIncrement>0x1</dimIncrement>
              <dimIndex>0-31</dimIndex>
              <name>S%s</name>
              <description>S00</description>
              <bitOffset>0</bitOffset>
              <bitWidth>1</bitWidth>
            </field>
          </fields>
        </register>
```
After patch it fails on descriptions.
But real issue is it should produce ` <dimIndex>00,01,02,..,31</dimIndex>` instead of `<dimIndex>0-31</dimIndex>`.

**Comment by kossnikita** [comment]
Looks good now

```xml
<field>
  <dim>32</dim>
  <dimIncrement>0x1</dimIncrement>
  <dimIndex>00,01,02,03,04,05,06,07,08,09,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31</dimIndex>
  <name>S%s</name>
  <description>S%s</description>
  <bitOffset>0</bitOffset>
  <bitWidth>1</bitWidth>
</field>
```

**Comment by kossnikita** [comment]
I have problem with this but I don't understand the problem

```yaml
  _array:
    DT?,DT[1][0]:
      description: Battery powered domain data register %s
    DT??:
      description: Battery powered domain data register %s
```

```xml
<registers>
    <register>
      <name>DT1</name>
      <displayName>DT1</displayName>
      <description>Battery powered domain data
      register (BPR_DTx)</description>
    </register>
    <register>
      <name>DT2</name>
      <displayName>DT2</displayName>
      <description>Battery powered domain data
      register (BPR_DTx)</description>
    </register>
    <register>
      <name>DT3</name>
      <displayName>DT3</displayName>
      <description>Battery powered domain data
      register (BPR_DTx)</description>
    </register>
    <register>
      <name>DT4</name>
      <displayName>DT4</displayName>
      <description>Battery powered domain data
      register (BPR_DTx)</description>
    </register>
    <register>
      <name>DT5</name>
      <displayName>DT5</displayName>
      <description>Battery powered domain data
      register (BPR_DTx)</description>
    </register>
    <register>
      <name>DT6</name>
      <displayName>DT6</displayName>
      <description>Battery powered domain data
      register (BPR_DTx)</description>
    </register>
    <register>
      <name>DT7</name>
      <displayName>DT7</displayName>
      <description>Battery powered domain data
      register (BPR_DTx)</description>
    </register>
    <register>
      <name>DT8</name>
      <displayName>DT8</displayName>
      <description>Battery powered domain data
      register (BPR_DTx)</description>
    </register>
    <register>
      <name>DT9</name>
      <displayName>DT9</displayName>
      <description>Battery powered domain data
      register (BPR_DTx)</description>
    </register>
    <register>
      <name>DT10</name>
      <displayName>DT10</displayName>
      <description>Battery powered domain data
      register (BPR_DTx)</description>
    </register>
    <register>
      <name>DT11</name>
      <displayName>DT11</displayName>
      <description>Battery powered domain data
      register (BPR_DTx)</description>
    </register>
...
</registers>
```

```
    Caused by:
        0: Processing device `AT32F403`
        1: According to `BPR`
        2: Processing peripheral `BPR`
        3: Collecting registers matched to `DT?,DT[1][0]` in array
        4: BPR: registers cannot be collected into DT?,DT[1][0] array. Please, specify displayName
```

File: src/patch/mod.rs [hunk 2]
```diff
@@ -637,3 +638,48 @@ fn check_offsets(offsets: &[u32], dim_increment: u32) -> bool {
...
+}
+/// Tries to get common description (or displayNames) for register/field array with "%s" in index position.
+/// Returns `None` if incoming descriptions have more then 1 difference
+fn common_description(descs: &[Option<&str>], dim_index: &[String]) -> Option<Option<String>> {
+    let mut start = 0;
+    if descs.len() > 1 {
+        if let Some((d1, d2)) = descs[0].zip(*descs.last().unwrap()) {
+            start = first_diff(d1, d2);
+        }
+    }
```
Comment (burrbull) [inline src/patch/mod.rs:668]: This is because of searching first difference.
Alternative is fully delete this part.

**Comment by burrbull** [comment]
cc @kossnikita , @Emilgardis looks like solution is founded

**Comment by burrbull** [comment]
let's merge this

**Comment by kossnikita** [comment]
Wait a little longer please. I haven't finished my tests yet.

**Comment by kossnikita** [comment]
LGTM. Could you update changelog?

**Comment by burrbull** [comment]
> LGTM. Could you update changelog?

Sure.

File: example/common_patches/tsc/tsc.yaml [hunk 3]
```diff
@@ -4,5 +4,5 @@ TSC:
   IOHCR:
     _array:
       G1_IO?: {}
-    G2_IO*:
-      _derivedFrom: "G1_IO1"
+    #G2_IO*:
+    #  _derivedFrom: "G1_IO1"
```
Comment (Emilgardis) [inline example/common_patches/tsc/tsc.yaml:8]: is this intentional?

**Review by Emilgardis — APPROVED** [review]

File: example/common_patches/tsc/tsc.yaml (hunk 3 above)
Comment (kossnikita) [reply example/common_patches/tsc/tsc.yaml:8]: I'm not sure what it is. This doesn't work right now and results in an error.

**Pull request merged**